java/test/com/google/jglome/GlomeTest.java

```diff
@@ -0,0 +1,177 @@
+  KeyPair[] aKeys = new KeyPair[N_TEST_VECTORS];
+  KeyPair[] bKeys = new KeyPair[N_TEST_VECTORS];
+
+  GlomeTest() throws MinPeerTagLengthOutOfBoundsException, InvalidKeyException {
+    for (int i = 0; i < N_TEST_VECTORS; i++) {
+      TestVector testVector = TEST_VECTORS.get(i);
+      aKeys[i] = new KeyPair(testVector.getKa(), testVector.getKah());
+      bKeys[i] = new KeyPair(testVector.getKb(), testVector.getKbh());
+      glomeManagers[i][0] = new GlomeBuilder(bKeys[i].getPublicKey(), 32)
+          .setPrivateKey(aKeys[i].getPrivateKey())
+          .build();
+      glomeManagers[i][1] = new GlomeBuilder(aKeys[i].getPublicKey(), 28)
+          .setPrivateKey(bKeys[i].getPrivateKey())
+          .build();
+    }
+  }
+
+  @Test
+  public void testShouldFailWhenMinPeerTagLengthIsOutOfBounds() {
+    int[] minPeerTagLength = new int[]{MIN_TAG_LENGTH - 1, MAX_TAG_LENGTH + 1};
+
+    for (int len : minPeerTagLength) {
+      try {
+        new GlomeBuilder(aKeys[0].getPublicKey(), len);
+      } catch (MinPeerTagLengthOutOfBoundsException e) {
+        assertEquals(e.getMessage(),
+            String.format("minPeerTagLength argument should be in [%d..%d] range. Got %d.",
+                MIN_TAG_LENGTH, MAX_TAG_LENGTH, len));
+      }
+    }
+  }
+
+  @Test
+  public void checkCorrectMinPeerTagLength() {
+    for (int len = MIN_TAG_LENGTH; len <= MAX_TAG_LENGTH; len++) {
+      try {
+        new GlomeBuilder(aKeys[0].getPublicKey(), len);
+      } catch (MinPeerTagLengthOutOfBoundsException e) {
+        assertEquals(e.getMessage(),
+            String.format("minPeerTagLength argument should be in [%d..%d] range. Got %d.",
+                MIN_TAG_LENGTH, MAX_TAG_LENGTH, len));
+      }
+    }
+  }
+
+  @Test
+  public void testShouldFailWhenCounterIsOutOfBounds() {
+    TestVector vector = TEST_VECTORS.get(0);
+    int[] counters = new int[]{MIN_CNT_VALUE - 1, MAX_CNT_VALUE + 1};
+
+    for (int cnt : counters) {
+      try {
+        glomeManagers[0][0].generateTag(vector.getMsg(), cnt);
+      } catch (CounterOutOfBoundsException e) {
+        assertEquals(e.getMessage(),
+            String.format("Counter should be in [%d..%d] range. Got %d.", MIN_CNT_VALUE,
+                MAX_CNT_VALUE, cnt));
+      }
+    }
+  }
+
+  @Test
+  public void checkCorrectCounters() throws CounterOutOfBoundsException {
+    TestVector vector = TEST_VECTORS.get(0);
+
+    for (int cnt = MIN_CNT_VALUE; cnt < MAX_CNT_VALUE; cnt++) {
+      glomeManagers[0][0].generateTag(vector.getMsg(), cnt);
+    }
+  }
+
+  @Test
+  public void derivedKeyShouldEqualOriginalKey() {
+    for (int i = 0; i < N_TEST_VECTORS; i++) {
+      assertArrayEquals(aKeys[i].getPublicKey(), glomeManagers[i][0].userPublicKey());
+      assertArrayEquals(bKeys[i].getPublicKey(), glomeManagers[i][1].userPublicKey());
+    }
+  }
+
+  @Test
+  public void testTagGeneration() throws CounterOutOfBoundsException {
+    for (int i = 0; i < N_TEST_VECTORS; i++) {
```

Comment:
A general remark about for loops in tests: if I were to debug this test, it would be very helpful if there was an indication of the loop variable in the assertion error.  Especially true for loops with many iterations, but even for the two sets of test vectors it makes a difference, I'd say.

Reply:
What do you mean by an indication there? Is it mostly about naming and including it in the Error/Exception messages like `String.format("Test failed at step %d...", i)`?

Reply:
Yeah, is it possible to include some string with the assertion failure in Java? If so, that'd be the way to go here.

Reply:
Yes, it's possible. Added.